aea/helpers/async_friendly_queue.py

```diff
@@ -0,0 +1,83 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2018-2019 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+"""This module contains the implementation of AsyncFriendlyQueue."""
+import asyncio
+import queue
+from collections import deque
+from typing import Any
+
+
+class AsyncFriendlyQueue(queue.Queue):
+    """queue.Queue with async_get method."""
+
+    def __init__(self, *args, **kwargs):
+        """Init queue."""
+        super().__init__(*args, **kwargs)
+        self._non_empty_waiters = deque()
+
+    def put(self, item: Any, *args, **kwargs) -> None:
+        """
+        Put an item into the queue.
+
+        :param item: item to put in the queue
+        :param args, kwargs: similar to queue.Queue.put
+        """
+        super().put(item, *args, **kwargs)
```

Comment:
maybe we can do something slightly better (correct me if I'm mistaken). Basically, we want always to wake up only one thread that is waiting for messages.

With the current implementation, it seems to me we always wake up a coroutine, if any. But it might happen that there were already other threads waiting, and `super().put()` wakes up only one of them.

If that's the case, then we don't need to wake up any coroutine, because the waken thread will consume the message. This could save some time at scale.

That being said, I'm happy with the current implementation, and extra-work might be an overkill today. Thoughts?

Reply:
It's async friendly queue and the main idea is to wake up coroutine on the item put.
It assumed to be used with coroutines not threads.
I see no reason to create a solution for one consumer only.

Reply:
> It assumed to be used with coroutines not threads.

You mean only coroutines can consume the items in the queue?
However, that's not the case since the parent `get()` is still callable (it just forwards the call to super().get()`)

Reply:
I mean async_get created to support coroutines as well, queue can be consumed by get.
the main idea is to create quue to be consumable by synchronous and asynchronous code but not the same time.

Reply:
I agree solution I provided is not perfect, but the main problem is to create queue thread safe and consumable by coroutines running in a different thread

Reply:
@marcofavorito probably it's better to use  https://github.com/aio-libs/janus

Reply:
Yes, janus is an interesting project. Anyway as I said I'm happy with the current implementation and we can iterate over this if needed